amr-wind/equation_systems/icns/source_terms/DragForcing.cpp

```diff
@@ -0,0 +1,157 @@
+#include "amr-wind/equation_systems/icns/source_terms/DragForcing.H"
+#include "amr-wind/utilities/IOManager.H"
+
+#include "AMReX_Gpu.H"
+#include "AMReX_Random.H"
+#include "amr-wind/wind_energy/ABL.H"
+
+namespace amr_wind::pde::icns {
+
+DragForcing::DragForcing(const CFDSim& sim)
+    : m_sim(sim)
+    , m_mesh(sim.mesh())
+    , m_velocity(sim.repo().get_field("velocity"))
+{
+    amrex::ParmParse pp("DragForcing");
+    pp.query("drag_coefficient", m_drag);
+    pp.query("sponge_strength", m_sponge_strength);
+    pp.query("sponge_density", m_sponge_density);
+    pp.query("sponge_distanceX", m_sponge_distanceX);
+    pp.query("sponge_distanceY", m_sponge_distanceY);
+    const auto& phy_mgr = m_sim.physics_manager();
+    if (phy_mgr.contains("ABL")) {
+        const auto& abl = m_sim.physics_manager().get<amr_wind::ABL>();
+        const VelPlaneAveraging& fa_velocity =
+            abl.abl_statistics().vel_profile_coarse();
+        device_vel_ht.resize(fa_velocity.line_centroids().size());
+        device_vel_vals.resize(fa_velocity.line_average().size());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, fa_velocity.line_centroids().begin(),
+            fa_velocity.line_centroids().end(), device_vel_ht.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, fa_velocity.line_average().begin(),
+            fa_velocity.line_average().end(), device_vel_vals.begin());
+    } else {
+        m_sponge_strength = 0.0;
+    }
+}
+
+DragForcing::~DragForcing() = default;
+
+void DragForcing::operator()(
+    const int lev,
+    const amrex::MFIter& mfi,
+    const amrex::Box& bx,
+    const FieldState fstate,
+    const amrex::Array4<amrex::Real>& src_term) const
+{
+    const auto& vel =
+        m_velocity.state(field_impl::dof_state(fstate))(lev).const_array(mfi);
+    const bool is_terrain = this->m_sim.repo().int_field_exists("terrainBlank");
+    if (!is_terrain) {
+        amrex::Abort("Need terrain blanking variable to use this source term");
+    }
+    auto* const m_terrain_blank =
+        &this->m_sim.repo().get_int_field("terrainBlank");
+    const auto& blank = (*m_terrain_blank)(lev).const_array(mfi);
+    auto* const m_terrain_drag =
+        &this->m_sim.repo().get_int_field("terrainDrag");
+    const auto& drag = (*m_terrain_drag)(lev).const_array(mfi);
+    auto* const m_terrainz0 = &this->m_sim.repo().get_field("terrainz0");
+    const auto& terrainz0 = (*m_terrainz0)(lev).const_array(mfi);
+    const auto& geom_vec = m_mesh.Geom();
+    const auto& geom = geom_vec[lev];
+    const auto& dx = geom.CellSize();
+    const auto& prob_lo = geom.ProbLoArray();
+    const auto& prob_hi = geom.ProbHiArray();
+    const amrex::Real device_drag = m_drag;
```

Comment:
sorry to be a pain. But you can remove all the `device_` prefix here. That's the nice thing about using the `m_` pfx in the first place. There's a bunch of places like that here.